docs/docs/dev_docs/contracts/syntax/storage.md

```diff
@@ -1,40 +1,533 @@
+
+### `assert_contains_and_remove`
+
+This function is used to check existence of a note and then remove it without having read the note ahead of time. This can be useful for cases where the user would is providing all the information needed, such as cases where the note was never emitted to the network and thereby available to the wallet.
+
+#include_code assert_contains_and_remove /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust
+
+<!---
+@LHerskind
+I don't see why this one is actually needed and could not be deprecated down the line. 
+Allow insertions into the rpc database and this one can be removed.
+-->
+
+### `assert_contains_and_remove_publicly_created`
+
+Like above, this is used to ensure that the message exists in the data tree and then consume it. However, it differs slightly since there is currently a difference between notes that have been inserted from public and private execution. This means that you currently must use this function to consume and nullify a note that was created in a public function. This will be fixed in the future. 
+
+#include_code assert_contains_and_remove_publicly_created /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust
+
+While this might look intimidating, the use of the function is rather easy, and is used in the following way:
+
+#include_code assert_contains_and_remove_publicly_created /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust
+
+The reason we are not reading this note ahead of time is that no [encrypted log](./events.md#encrypted-events) was emitted for this note, since it was created in public thereby making the encrypted log useless (everyone saw the content ahead of time).
+
+### `remove`
+
+Will remove a note from the set if it previously have been read from storage. That being that you have fetched it through a `get_notes` call. This is useful when you want to remove a note that you have previously read from storage to not have to read it again. If you recall from earlier, we are emitting a nullifier when reading values to make sure that they are up to date. 
+
+#include_code remove /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust
+
+An example of how to use this operation is visible in the `easy_private_state`:
+
+#include_code remove /yarn-project/aztec-nr/easy-private-state/src/easy_private_state.nr rust
+
+### `get_notes`
+
+This function returns the notes the account has access to:
+
+#include_code get_notes /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust
+
+There's a limit on the maximum number of notes this function can return at a time. Check [here](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/aztec-nr/aztec/src/constants_gen.nr) and look for `MAX_READ_REQUESTS_PER_CALL` for the up-to-date number.
+
+Because of this limit, we should always consider using the second argument `NoteGetterOptions` to target the notes we need, and to reduce the time required to recursively call this function.
```

Comment:
What does it mean to "to reduce the time required to recursively call this function."? Can you say a bit more about why this is important?

Reply:
Will extend. 

The TL;DR:
When you are reading, you spend read requests for every call that you need to constrain. So the idea is that you find your notes, and then you apply a filter to those, such that only a subset is require to be read -> minimizing the number of calls you need to make and thereby the "time required to recursive call this function".

As an example, say that you have a lot of notes for a specific asset. Say you got 20 notes or something of different size. You want to transfer 10 tokens, which you can do with just a few of your tokens. You then use the filter to "stop" making more reads as soon as you found enough to perform the transfer. This could mean that you need to do just a few reads instead of reading all 20 notes, which could be the difference between using 1 second in proving or 10 seconds.